src/p11_ec.c

```diff
@@ -251,6 +255,30 @@ static EC_KEY *pkcs11_get_ec(PKCS11_KEY *key)
 	if (no_point && key->isPrivate) /* Retry with the public key */
 		no_point = pkcs11_get_point(ec, pkcs11_find_key_from_key(key));
 
+	/* Retry with the certificate */
+	if (no_point && key->isPrivate) {
+		cert = pkcs11_find_certificate(key);
```

Comment:
This returns a certificate which has the same CKA_ID as the key, doesn't it? In an ideal world that might match... but we shouldn't assume that. By all means look for such a cert as a best effort, but let's actually do a test operation and *check* that it matches, before filling it in.

Fixing OpenSSL would also be useful.

Reply:
> This returns a certificate which has the same CKA_ID as the key, doesn't it? In an ideal world that might match... but we shouldn't assume that. By all means look for such a cert as a best effort, but let's actually do a test operation and _check_ that it matches, before filling it in.

Yes, it gets a certificate with matching CKA_ID. I though this would be acceptable since the same is done when searching for matching public key objects. As you pointed out, this is a best effort to improve the current situation.

If we choose to check if the public keys match before filling the structure, we should do the same for public key objects, and for other algorithms.

> 
> Fixing OpenSSL would also be useful.

After reading about your last attempt to fix it there, I feel discouraged to do so.

Reply:
@dwmw2 Do you have a suggestion on how to check if the private key and a public key match for EC keys without performing a signature/verification?

For most of devices, doing an extra signature every time a key is used is impractical, since it can take too much time. I would suggest to check if the groups of the private key and the public key match and if the public key (point) is on the curve. But that would only guarantee that both  are using the same parameters.